### ISSUE TITLE
e2e gke flake: `Failed waiting for pods to be running` in `should be able to create a functioning NodePort service` test

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#20774]
Merge queue currently blocked on this failure: 
```
• Failure [126.005 seconds]
Services
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:862
  should be able to create a functioning NodePort service [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:415

  Feb  6 05:39:50.717: Failed waiting for pods to be running: Timeout waiting for %!d(MISSING) pods to be ready

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:1702
-
```

https://console.developers.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gke/513/

(Submitted a PR to fix the format string `MISSING` as #20773 but that's not the cause here)

**Comment by justinsb** [comment]
So the failure was
```
Feb  6 05:39:50.717: INFO: Pod nodeport-test-v06d8 failed to be running and ready.
```

The scheduler assigned it to `gke-jkns-gke-e2e-ci-36cd78e7-node-zk5x`:
 ```
Feb  6 05:39:50.721: INFO: event for nodeport-test-v06d8: {default-scheduler } Scheduled: Successfully assigned nodeport-test-v06d8 to gke-jkns-gke-e2e-ci-36cd78e7-node-zk5x
```

But that was a poor choice; multiple pods running on that node are stuck in pending (even ones unrelated to this test):

```
...
Feb  6 05:39:50.741: INFO: rand-non-local-bigpk                                          gke-jkns-gke-e2e-ci-36cd78e7-node-zk5x  Pending         [{Read
y False 0001-01-01 00:00:00 +0000 UTC 2016-02-06 05:38:18 -0800 PST ContainersNotReady containers with unready status: [c]}]
Feb  6 05:39:50.741: INFO: rand-non-local-cyin6                                          gke-jkns-gke-e2e-ci-36cd78e7-node-zk5x  Pending         [{Read
y False 0001-01-01 00:00:00 +0000 UTC 2016-02-06 05:38:18 -0800 PST ContainersNotReady containers with unready status: [c]}]
Feb  6 05:39:50.741: INFO: update-demo-nautilus-80q5v                                    gke-jkns-gke-e2e-ci-36cd78e7-node-zk5x  Pending         [{Read
y False 0001-01-01 00:00:00 +0000 UTC 2016-02-06 05:37:43 -0800 PST ContainersNotReady containers with unready status: [update-demo]}]
...
Feb  6 05:39:50.741: INFO: nodeport-test-v06d8                                           gke-jkns-gke-e2e-ci-36cd78e7-node-zk5x  Pending         [{Ready False 0001-01-01 00:00:00 +0000 UTC 2016-02-06 05:37:48 -0800 PST ContainersNotReady containers with unready status: [netexec]}]
...
Feb  6 05:39:50.741: INFO: kube-proxy-gke-jkns-gke-e2e-ci-36cd78e7-node-zk5x             gke-jkns-gke-e2e-ci-36cd78e7-node-zk5x  Running         [{Ready True 0001-01-01 00:00:00 +0000 UTC 2016-02-06 05:36:13 -0800 PST  }]
...
```

(pods on other nodes are happily in the Running state)

**Comment by justinsb** [comment]
Good news though, we have the kubelet log from the node (`130.211.177.118:22-kubelet.log`):

There are a number of surprising things in that log:

Disk mount failure?
```
E0206 13:39:14.395135    3519 manager.go:1803] Error running pod "frontend-mbdgk_e2e-tests-kubectl-4ff1d(cbae897c-ccd6-11e5-a0e9-42010af00002)" container "php-redis": GenerateRunContainerOptions: impossible: cannot find the mounted volumes for pod "frontend-mbdgk_e2e-tests-kubectl-4ff1d(cbae897c-ccd6-11e5-a0e9-42010af00002)"
E0206 13:39:14.395204    3519 pod_workers.go:129] Error syncing pod cbae897c-ccd6-11e5-a0e9-42010af00002, skipping: failed to "StartContainer" for "php-redis" with RunContainerError: "GenerateRunContainerOptions: impossible: cannot find the mounted volumes for pod \"frontend-mbdgk_e2e-tests-kubectl-4ff1d(cbae897c-ccd6-11e5-a0e9-42010af00002)\""
```

Error reading metadata (or is this a normal 'no special docker config' message?):
```
I0206 13:39:14.401740    3519 config.go:155] body of failing http response: &{0xc208cf60c0 {0 0} false <nil> 0x5e11a0 0x5e1130}
E0206 13:39:14.401789    3519 metadata.go:124] while reading 'google-dockercfg-url' metadata: http status code: 404 while fetching url http://metadata.google.internal./computeMetadata/v1/instance/attributes/google-dockercfg-url
```

Bad container ids (normal?).  I've seen this on AWS before 20395 went in:
```
E0206 13:38:01.293536    3519 runtime.go:159] invalid container ID: ""
E0206 13:38:01.306289    3519 runtime.go:159] invalid container ID: ""
E0206 13:38:03.454731    3519 runtime.go:159] invalid container ID: ""
I0206 13:38:04.243079    3519 server.go:1059] GET /healthz: (35.937Âµs) 200 [[Go 1.1 package http] 127.0.0.1:48792]
```

Failed to create volume when launching our pod?
```
E0206 13:40:00.974080    3519 manager.go:1803] Error running pod "nodeport-test-v06d8_e2e-tests-services-r3djt(cfb3b0eb-ccd6-11e5-a0e9-42010af00002)" container "netexec": GenerateRunContainerOptions: impossible: cannot find the mounted volumes for pod "nodeport-test-v06d8_e2e-tests-services-r3djt(cfb3b0eb-ccd6-11e5-a0e9-42010af00002)"
```


I think the root cause is likely those "impossible: cannot find the mounted volumes" messages, but someone more familiar with kubelet should probably have a look at https://console.developers.google.com/m/cloudstorage/b/kubernetes-jenkins/o/logs/kubernetes-e2e-gke/513/artifacts/130.211.177.118%3A22-kubelet.log

**Comment by lavalamp** [comment]
@justinsb thanks for doing my job for me :)

**Comment by dchen1107** [comment]
@yujuhong Can you take a look at this one? Thanks!

**Comment by yujuhong** [comment]
Again, GKE logs are not very helpful (#20661).

@justinsb, thanks for the digging. AFAICT, those errors are mostly red herrings. kubelet wasn't able to create/start the container in time. The tests timed out and deleted the pod. 
So by the time kubelet wanted to start the container, the volumes were already gone.



#20781 had a similar problem, and the relevant logs are:
```
I0207 04:04:21.578006    3515 kubelet.go:2316] SyncLoop (ADD, "api"): "nodeport-test-wt9y2_e2e-tests-services-ga4sh(de019055-cd4f-11e5-98b0-42010af00007)"
W0207 04:04:21.578865    3515 manager.go:173] Container readiness changed for unknown container: "nodeport-test-wt9y2_e2e-tests-services-ga4sh(de019055-cd4f-11e5-98b0-42010af00007)" - "://"
I0207 04:04:21.591854    3515 manager.go:1551] Need to restart pod infra container for "nodeport-test-wt9y2_e2e-tests-services-ga4sh(de019055-cd4f-11e5-98b0-42010af00007)" because it is not found
I0207 04:04:22.149903    3515 kubelet.go:2338] SyncLoop (PLEG): "nodeport-test-wt9y2_e2e-tests-services-ga4sh(de019055-cd4f-11e5-98b0-42010af00007)", event: &pleg.PodLifecycleEvent{ID:"de019055-cd4f-11e5-98b0-42010af00007", Type:"ContainerStarted", Data:"5fca6b5e43a03a31739abd73841af77e61ba687a16368418871ff178a13d8057"}
---- test timed out and deleted the pod ----
I0207 04:06:25.161553    3515 kubelet.go:2322] SyncLoop (REMOVE, "api"): "nodeport-test-wt9y2_e2e-tests-services-ga4sh(de019055-cd4f-11e5-98b0-42010af00007)"
I0207 04:06:25.164599    3515 kubelet.go:2157] Killing unwanted pod "nodeport-test-wt9y2"
I0207 04:06:25.171678    3515 manager.go:1247] Killing container "5fca6b5e43a03a31739abd73841af77e61ba687a16368418871ff178a13d8057 e2e-tests-services-ga4sh/nodeport-test-wt9y2" with 0 second grace period
I0207 04:06:25.224834    3515 kubelet.go:1887] volume "de019055-cd4f-11e5-98b0-42010af00007/default-token-4efrg", still has a container running "de019055-cd4f-11e5-98b0-42010af00007", skipping teardown
I0207 04:06:25.528278    3515 manager.go:1279] Container "5fca6b5e43a03a31739abd73841af77e61ba687a16368418871ff178a13d8057 e2e-tests-services-ga4sh/nodeport-test-wt9y2" exited after 356.547366ms
W0207 04:06:27.225562    3515 kubelet.go:1892] Orphaned volume "de019055-cd4f-11e5-98b0-42010af00007/default-token-4efrg" found, tearing down volume
E0207 04:06:27.529492    3515 manager.go:1803] Error running pod "nodeport-test-wt9y2_e2e-tests-services-ga4sh(de019055-cd4f-11e5-98b0-42010af00007)" container "netexec": GenerateRunContainerOptions: impossible: cannot find the mounted volumes for pod "nodeport-test-wt9y2_e2e-tests-services-ga4sh(de019055-cd4f-11e5-98b0-42010
af00007)"
E0207 04:06:27.529558    3515 pod_workers.go:129] Error syncing pod de019055-cd4f-11e5-98b0-42010af00007, skipping: failed to "StartContainer" for "netexec" with RunContainerError: "GenerateRunContainerOptions: impossible: cannot find the mounted volumes for pod \"nodeport-test-wt9y2_e2e-tests-services-ga4sh(de019055-cd4f-11e
5-98b0-42010af00007)\""
E0207 04:06:27.535807    3515 event.go:193] Server rejected event '&api.Event{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:api.ObjectMeta{Name:"nodeport-test-wt9y2.14308d406f8a3bb8", GenerateName:"", Namespace:"e2e-tests-services-ga4sh", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:unversioned.Time{Time:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*unversioned.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil)}, InvolvedObject:api.ObjectReference{Kind:"Pod", Namespace:"e2e-tests-services-ga4sh", Name:"nodeport-test-wt9y2", UID:"de019055-cd4f-11e5-98b0-42010af00007", APIVersion:"v1", ResourceVersion:"2445", FieldPath:"spec.containers{netexec}"}, Reason:"Pulled", Message:"Successfully pulled image \"gcr.io/google_containers/netexec:1.4\"", Source:api.EventSource{Component:"kubelet", Host:"gke-jkns-gke-e2e-ci-1072ced8-node-0sd7"}, FirstTimestamp:unversioned.Time{Time:time.Time{sec:63590414787, nsec:529423800, loc:(*time.Location)(0x2d76ba0)}}, LastTimestamp:unversioned.Time{Time:time.Time{sec:63590414787, nsec:529423800, loc:(*time.Location)(0x2d76ba0)}}, Count:1, Type:"Normal"}': 'namespaces "e2e-tests-services-ga4sh" not found' (will not retry!)
E0207 04:06:27.538187    3515 event.go:193] Server rejected event '&api.Event{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:api.ObjectMeta{Name:"nodeport-test-wt9y2.14308d406f8c8d99", GenerateName:"", Namespace:"e2e-tests-services-ga4sh", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:unversioned.Time{Time:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*unversioned.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil)}, InvolvedObject:api.ObjectReference{Kind:"Pod", Namespace:"e2e-tests-services-ga4sh", Name:"nodeport-test-wt9y2", UID:"de019055-cd4f-11e5-98b0-42010af00007", APIVersion:"v1", ResourceVersion:"2445", FieldPath:""}, Reason:"FailedSync", Message:"Error syncing pod, skipping: failed to \"StartContainer\" for \"netexec\" with RunContainerError: \"GenerateRunContainerOptions: impossible: cannot find the mounted volumes for pod \\\"nodeport-test-wt9y2_e2e-tests-services-ga4sh(de019055-cd4f-11e5-98b0-42010af00007)\\\"\"\n", Source:api.EventSource{Component:"kubelet", Host:"gke-jkns-gke-e2e-ci-1072ced8-node-0sd7"}, FirstTimestamp:unversioned.Time{Time:time.Time{sec:63590414787, nsec:529575833, loc:(*time.Location)(0x2d76ba0)}}, LastTimestamp:unversioned.Time{Time:time.Time{sec:63590414787, nsec:529575833, loc:(*time.Location)(0x2d76ba0)}}, Count:1, Type:"Warning"}': 'namespaces "e2e-tests-services-ga4sh" not found' (will not retry!)
```

At  04:06:27.538187, kubelet attempted to send an event *after* successfully pulling image but the pod was already deleted. It's most likely that kubelet was stuck at image pulling for a long time.
As for *why* kubelet pulled image so slowly, I think it may be related to the issue I just filed (#20836). Some tests use non-tagged images (e.g. busybox), which caused kubelet to always pull an image even though they are on the node. This may have caused a cascading effects of slow image pulls.

**Comment by dchen1107** [comment]
Should this one closed via #20837?

**Comment by yujuhong** [comment]
I think #20837 should greatly reduce the chance of this happening. That said, we still have quite a lot of untagged images used in the tests, so it's not 100% fixed. I'll keep this one open for another day and see if there are any more flakes.

**Comment by dchen1107** [comment]
I am closing this one. Let's fix other image issue when hitting them.

**Comment by jlowdermilk** [comment]
This started flaking again last night.
[jenkins history](http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gke/2214/testReport/junit/(root)/Kubernetes%20e2e%20suite/Services_should_be_able_to_create_a_functioning_NodePort_service/history/)

[kubernetes-e2e-gke/2205](https://storage.cloud.google.com/kubernetes-jenkins/logs/kubernetes-e2e-gke/2205)
[kubernetes-e2e-gke/2211](https://storage.cloud.google.com/kubernetes-jenkins/logs/kubernetes-e2e-gke/2211)
[kubernetes-e2e-gke/2213](https://storage.cloud.google.com/kubernetes-jenkins/logs/kubernetes-e2e-gke/2213)
kubelet and docker logs should be in the bucket.

**Comment by yujuhong** [comment]
From http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gke/2215 (https://cloud.google.com/console/storage/kubernetes-jenkins/logs/kubernetes-e2e-gke/2215):

```
09:55:11   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework.go:101
09:55:11 STEP: Collecting events from namespace "e2e-tests-services-ypt1t".
09:55:11 Mar  4 09:54:54.773: INFO: At 2016-03-04 09:52:52 -0800 PST - event for nodeport-test: {replication-controller } SuccessfulCreate: Created pod: nodeport-test-xm6wf
09:55:11 Mar  4 09:54:54.773: INFO: At 2016-03-04 09:52:55 -0800 PST - event for nodeport-test-xm6wf: {default-scheduler } Scheduled: Successfully assigned nodeport-test-xm6wf to gke-jenkins-e2e-e5eb611b-node-2f9b
09:55:11 Mar  4 09:54:54.773: INFO: At 2016-03-04 09:53:33 -0800 PST - event for nodeport-test-xm6wf: {kubelet gke-jenkins-e2e-e5eb611b-node-2f9b} Pulling: pulling image "gcr.io/google_containers/netexec:1.4"
09:55:11 Mar  4 09:54:54.773: INFO: At 2016-03-04 09:53:33 -0800 PST - event for nodeport-test-xm6wf: {kubelet gke-jenkins-e2e-e5eb611b-node-2f9b} Failed: Failed to pull image "gcr.io/google_containers/netexec:1.4": API error (500): unable to ping registry endpoint https://gcr.io/v0/
09:55:11 v2 ping attempt failed with error: Get https://gcr.io/v2/: dial tcp: lookup gcr.io on 10.240.0.1:53: dial udp 10.240.0.1:53: network is unreachable
09:55:11  v1 ping attempt failed with error: Get https://gcr.io/v1/_ping: dial tcp: lookup gcr.io on 10.240.0.1:53: dial udp 10.240.0.1:53: network is unreachable
09:55:11 
09:55:11 Mar  4 09:54:54.773: INFO: At 2016-03-04 09:53:33 -0800 PST - event for nodeport-test-xm6wf: {kubelet gke-jenkins-e2e-e5eb611b-node-2f9b} FailedSync: Error syncing pod, skipping: failed to "StartContainer" for "netexec" with ErrImagePull: "API error (500): unable to ping registry endpoint https://gcr.io/v0/\nv2 ping attempt failed with error: Get https://gcr.io/v2/: dial tcp: lookup gcr.io on 10.240.0.1:53: dial udp 10.240.0.1:53: network is unreachable\n v1 ping attempt failed with error: Get https://gcr.io/v1/_ping: dial tcp: lookup gcr.io on 10.240.0.1:53: dial udp 10.240.0.1:53: network is unreachable\n"
09:55:11 
09:55:11 Mar  4 09:54:54.774: INFO: At 2016-03-04 09:53:34 -0800 PST - event for nodeport-test-xm6wf: {kubelet gke-jenkins-e2e-e5eb611b-node-2f9b} BackOff: Back-off pulling image "gcr.io/google_containers/netexec:1.4"
09:55:11 Mar  4 09:54:54.774: INFO: At 2016-03-04 09:53:34 -0800 PST - event for nodeport-test-xm6wf: {kubelet gke-jenkins-e2e-e5eb611b-node-2f9b} FailedSync: Error syncing pod, skipping: failed to "StartContainer" for "netexec" with ImagePullBackOff: "Back-off pulling image \"gcr.io/google_containers/netexec:1.4\""
```

I checked a few other builds, and they all seem to be related to slow image pulling (gcr.io issues).

Related issues: #20552, #20550

/cc @ArtfulCoder

**Comment by yujuhong** [comment]
The revert of #22555 is in progress.

In short, I added retry/timeout in ReadUrl(), which reads the metadata url so that we have more clarity when it fails to do so.

However, the same function is used to read [medadata keys](https://github.com/kubernetes/kubernetes/blob/9d3806bceb5097793f2d7ccf60427c4a2756372a/pkg/credentialprovider/gcp/metadata.go#L32) (`google-dockercfg`, `google-dockercfg-url`) that don't exist in our test cluster. Whenever there is a image pulling request, we check the cache to see if the credentials have expired (60s). If so, the code will then try to refresh the cache by accessing every URL it knows. Since these keys don't exist, we ended up spending the majority of time retrying until timeout for each key before pulling the image. The revert should help restore things to what they were before.

The registry ping failure in https://github.com/kubernetes/kubernetes/issues/20774#issuecomment-192395062 is a separate issue though.
